[PSR-2] Switch case space fixer

Soullivaneuh opened the pull request:
Closes #1542.

@keradus readme generator seems to be unsynced on `1.11` (elements are removed and my fixer not added). Could you please have a look?

SpacePossum commented:
please take a look @ https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1539#issuecomment-159541538 for some more (`horror`) ways of defining a `case` in a switch ;)

Soullivaneuh commented:
@SpacePossum I don't know why should I do with comment. Remove it? keep it?

Soullivaneuh commented on Symfony/CS/Tests/Fixer/PSR2/SwitchCaseSpaceFixerTest.php:
```diff
@@ -0,0 +1,173 @@
+                    case /**/(/**/ // test
+                        4
+                        /**/)//
+                        /**/ : {
```
@SpacePossum @keradus serisouly, what should I do with this sh*t?! x)

Soullivaneuh replied:
For the moment, the fixer just remove space between the last comment (`/**/`) and `:`.

gharlan replied:
> For the moment, the fixer just remove spaces between the last comment (/**/) and :.

:+1:

SpacePossum replied:
sounds good to me, it is what the fixer should do :)
(I opt the tests to make sure the fixer does not delete the comments and does not crash, so all good to me)

gharlan commented:
IMO the fixer should also handle the `default` case.
What do you think?

keradus commented:
> @keradus readme generator seems to be unsynced on 1.11 (elements are removed and my fixer not added). Could you please have a look?

? readme file on 1.11 is good

keradus commented on Symfony/CS/Fixer/PSR2/SwitchCaseSpaceFixer.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * @author Sullivan Senechal <soullivaneuh@gmail.com>
+ */
+class SwitchCaseSpaceFixer extends AbstractFixer
```
final

keradus commented on Symfony/CS/Tests/Fixer/PSR2/SwitchCaseSpaceFixerTest.php:
```diff
@@ -0,0 +1,173 @@
+/**
+ * @author Sullivan Senechal <soullivaneuh@gmail.com>
+ */
+class SwitchCaseSpaceFixerTest extends AbstractFixerTestBase
```
final, internal

keradus commented:
readme IS ok, tests show that the readme generated from code has info about `switch_case_space` and readme in file has not. just regenerate the readme.

keradus commented on Symfony/CS/Fixer/PSR2/SwitchCaseSpaceFixer.php:
```diff
@@ -0,0 +1,41 @@
+
+        $caseTokens = $tokens->findGivenKind(T_CASE);
+
+        foreach ($caseTokens as $caseIndex => $caseToken) {
```
```php

foreach ($tokens as $index => $token) {
    if (!$token->isGivenKind(T_CASE)) {
        continue;
    }

    ...
}

Soullivaneuh replied:
Sorry but why this? 

Only `T_CASE` tokens is needed for this fixers, why not simply using `findGivenKind`?

SpacePossum replied:
I think is a good idea to check all tokens from end to start, because if you remove/clear or add tokens the indexes of other tokens may change. If you fetch indexes first and then manipulate the collection the indexes fetched may no longer be what you expect.

keradus replied:
Second reason is performance, as `findGivenKind` is one loop on whole collection, and then you do a second look on partial collection.

It would be great if `findGivenKind` would be a generator, but it is not.

SpacePossum commented:
>IMO the fixer should also handle the default case.
>What do you think?

:+1:

Soullivaneuh commented:
> readme IS ok, tests show that the readme generated from code has info about switch_case_space and readme in file has not. just regenerate the readme.

Yeah my bad, my unix shell was using the global one. --"

Soullivaneuh commented:
Code updated, PR ready for review.

keradus commented on Symfony/CS/Fixer/PSR2/SwitchCaseSpaceFixer.php:
```diff
@@ -0,0 +1,43 @@
+                continue;
+            }
+
+            $colonIndex = $tokens->getNextTokenOfKind($index, [':']);
```
please fix project itself

Soullivaneuh commented:
@keradus done.

gharlan commented:
test case:

```php
switch ($a) {
    case $b ? 'c' : 'd' :
        break;
}
```

Soullivaneuh commented:
@gharlan Well, I'm wondering how I can fix this...

gharlan commented:
I think you have to count the `?` and ignore same amount of `:`.

Soullivaneuh commented:
Yeah I was starting this way.

gharlan commented:
Another test case: `$b ?: $c`
(Should pass automatically when the other test case is fixed. Just want to be sure. ;))

Soullivaneuh commented:
Should be OK. :+1:

keradus commented:
Closing in favour of #1550

gharlan commented:
@keradus this is a different fixer (spaces instead of parentheses).

Soullivaneuh commented:
Yes @keradus, this one should not be closed.

gharlan commented:
@Soullivaneuh maybe he closed this because of your comment in #1550:

> Closes #1543 

;)

keradus commented:
Requested by you then closed ;)
OK, reopened.

Soullivaneuh commented:
Well, I have to stop drinking. :wine_glass:

Soullivaneuh commented:
Ready for review and merge.

Soullivaneuh commented:
Tests passed. Ready to merge for me.

keradus commented on Symfony/CS/Fixer/PSR2/SwitchCaseSpaceFixer.php:
```diff
@@ -0,0 +1,67 @@
+                }
+
+                if ($tokens[$colonIndex]->equals(':')) {
+                    --$questionMarkCount;
```
odd variable name for `:` ;)

Soullivaneuh replied:
Well... I... sh*t. :-)

keradus commented:
Could anyone make a review of logic please?

gharlan commented:
:+1: and thanks for implementing this!

Soullivaneuh commented:
@keradus and what about this one for 1.11?

keradus commented:
nope, 1.11 is out there!

Soullivaneuh commented:
![image](https://cloud.githubusercontent.com/assets/1698357/11516601/8d23dec6-9884-11e5-9437-9ea45986a772.png)

I squashed and updated for 1.12 then. :+1:

gharlan commented:
> :+1: and thanks for implementing this!

My :+1: is still valid (reviewed again).
Anyone else?

keradus commented on Symfony/CS/Fixer/PSR2/SwitchCaseSpaceFixer.php:
```diff
@@ -0,0 +1,67 @@
+            $valueIndex = $tokens->getPrevNonWhitespace($colonIndex);
+
+            if ($colonIndex - $valueIndex > 1) {
+                $tokens->clearRange($valueIndex + 1, $colonIndex - 1);
```
running your tests it is always true:
`$valueIndex + 1 === $colonIndex - 1`

not a big range I say...

Soullivaneuh replied:
The condition is to check if you have something to clear between tokens.

So with your style, it should be:

```php
$valueIndex + 1 !== $colonIndex - 1
```

No?

keradus replied:
I'm not talking about the style here. The problem is that in all the test cases you have provided, in run time the first argument is exactly the same as second one.
So you could simply wrote `$tokens[$valueIndex + 1]->clear();` and have the same effect.

I bet it is not what you want and just need to extend the tests

Soullivaneuh replied:
Not sure to understand well.

I added a test, tell me if this i what you want. If not, please give me one! :-)

keradus replied:
OK, let use some code ;)

https://github.com/Soullivaneuh/PHP-CS-Fixer/pull/3

Why does everything works with my change? If `clearRange` is not needed please use my PR. If it is needed proove it in tests.

Soullivaneuh replied:
Indeed.

Soullivaneuh commented:
commit added.

SpacePossum commented:
Test case;
```php
            "<?php
switch(1) {
    case 1 ? Foo::bla() : 'e'     :    {

    }
    case Foo::bla()      :    {

    }
}
            "
```
passes as well :)
 :+1: 
(please rebase)

keradus commented:
And squash your two commits ;)

SpacePossum commented:
ping @Soullivaneuh   :)

keradus commented:
> And squash your two commits ;)

I think this is the only thing that left to merge this PR?
ping @Soullivaneuh

Soullivaneuh commented:
@keradus squashed and rebased.

keradus commented:
I see that my commit is missing, but I don't remember what was that. Do you squash it into your commit or forgot to include ?

GrahamCampbell commented:
Yeh, you should have hard reset to your last commit, squashed and rebased, then cherry-picked the other commit ontop.

GrahamCampbell commented:
That way you'd end up with a PR that was squashed by author.

keradus commented:
On just use interactive rebase and set what to do with each commit ;)
One way or another, it was my commit and I don't need it to be show as "mine". What I need is to have its changes. Are they there?

Soullivaneuh commented:
> I see that my commit is missing, but I don't remember what was that. Do you squash it into your commit or forgot to include ?

Oh, I didn't pay attention and only done a `rebase -i HEAD~3`. :-/

Which commit is missing exactly?

keradus commented:
There were 2 yours and 1 mine, Cannot point the commit now that you get rid of, didn't remember its hash...

If you squash last 3 commits then you squash mine as well, so no changes are lost.

In future please squash carefully (and read carefully - I asked to squash 2 of yours, not all - if all commits would be yours I would squash them during merging).

-------------------------------------

OK, thanks for info.
:+1:

keradus commented:
Thank you @Soullivaneuh.
